### PR TITLE
Undo grouping

### LINKED ISSUE #4151: Undo button for completed grouping phase
## Issue - Enhancement

This issue defines AC resulting from the discussion #3701, "Allow for regrouping of cards until vote phase completed" addressing the following JTBD:

> When a user has accidentally advanced the meeting, or wishes to add new reflections, or wishes to divide or combine two groups after having advanced to the Vote phase, they want an easy way to Undo the completion state of the Group phase, so that they can adjust the meeting content to their liking

@ackernaut please let me know if this would be blocked be design, or if you think the below AC are sufficient

### Acceptance Criteria (optional)
A design for when users wish to to undo the completion status of the Retro Group phase:
   - The 'Phase Completed' badge of the Group base is updated to read 'Phase Completed Undo' for the Facilitator, where the word Undo is a clickable link
   - Undoing the phase allows the Grouping phase to continue
   - Any non-facilitator users navigated to the Vote phase are navigated backward to the Group phase

Plus:
   - Opened an implementation issue to implement the resulting design

- **Estimated effort:** 12 hours ([see CONTRIBUTING.md](https://github.com/ParabolInc/parabol/blob/master/CONTRIBUTING.md#points-and-sizes))

### PR DESCRIPTION
**Pull request by tiffanyhan**
#4151

**AC:**
- [x] after starting a new meeting, adding reflections, grouping them, optionally voting on them, and optionally adding discussion comments and tasks-- the facilitator can navigate back to the completed group phase and type in `ididntmeanto`
- [x] any votes on all reflection groups are removed
- [x] any discussion items, including comments, tasks, and replies, in the meeting are deleted
- [x] all discussion stages are reset so there's only 1 placeholder discussion stage in the discussion phase. 
- [x] all stages and phases after GROUP are reset (basically, the `Meeting` object looks identical to how it looked before the meeting was advanced from GROUP to VOTE)
- [x] as a facilitator, the group phase no longer looks completed
- [x] as a participants, they are transported back to the group phase

### CODE REVIEW

**Comment by mattkrick** [comment]
requesting review from @nickoferrall

File: packages/client/components/PhaseCompleted.tsx
Comment (nickoferrall) [inline packages/client/components/PhaseCompleted.tsx:17]: `canUndo` is always true at the moment and `Wrapper` in `UndoablePhaseCompleted` seems to be identical to the `PhaseCompleteTag` styled component. Perhaps we could remove this file, use `UndoablePhaseCompleted` in the parent and use the `PhaseCompleteTag` in `UndoablePhaseCompleted` to reduce duplication?

**Review by nickoferrall — APPROVED** [review]
LGTM! I love the cheat code. Once the merge conflicts are fixed, it looks good to go

File: packages/client/components/PhaseCompleted.tsx
Comment (tiffanyhan) [reply packages/client/components/PhaseCompleted.tsx:17]: @nickoferrall that is true! i just pushed a change to pass in `canUndo=true` only from `retroGroupPhase`. Once Terry has the design, `UndoablePhaseCompleted` should have different styling to include a clickable button to undo (instead of using the hotkey).  I thinkk it's good to hold off on further refactoring until the UI design is settled on.

File: packages/client/mutations/ResetMeetingToStageMutation.ts
Comment (tiffanyhan) [reply packages/client/mutations/ResetMeetingToStageMutation.ts:71]: good question! I was simply following the pattern of other mutations, and didn't look too much into what `safeProxy` does. looks like this is something that @mattkrick wrote, and my first guess is that it's a thin wrapper to ensure stricter type safety.

File: packages/client/mutations/ResetMeetingToStageMutation.ts
Comment (mattkrick) [inline packages/client/mutations/ResetMeetingToStageMutation.ts:77]: `viewerVoteCount` is already requested & relay should magically merge in the new value since the reflectionGroupIds match. try to comment out all these lines & see if it still works. if it does, we can get rid of this entire updater (yay less code!)

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/resetMeetingToStage.ts:45]: `foundResponse` is guaranteed to be truthy since the above line guarantees that it exists

File: packages/client/mutations/ResetMeetingToStageMutation.ts
Comment (tiffanyhan) [reply packages/client/mutations/ResetMeetingToStageMutation.ts:77]: yea i tried that and unfortunately relay doesn't automagically update the viewer vote count on reflection groups. was gonna ask you if you had any clues about why that would bwe.

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/resetMeetingToStage.ts:58]: you can grab this direct from the meeting & reduce a hit to the DB

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/resetMeetingToStage.ts:122]: nit: can cast to `RetroMeetingMember` to get free typechecking of the votesRemaining field

File: packages/client/mutations/ResetMeetingToStageMutation.ts
Comment (mattkrick) [reply packages/client/mutations/ResetMeetingToStageMutation.ts:77]: thaaat is interesting. let me pull a copy & find out why

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (tiffanyhan) [reply packages/server/graphql/mutations/resetMeetingToStage.ts:58]: ah wasn't aware we stored it on the meeting, thanks!

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (tiffanyhan) [reply packages/server/graphql/mutations/resetMeetingToStage.ts]: oh good catch

File: packages/client/components/PhaseCompleted.tsx
Comment (nickoferrall) [reply packages/client/components/PhaseCompleted.tsx:17]: Makes sense! Sounds good to me

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (mattkrick) [inline packages/server/graphql/mutations/resetMeetingToStage.ts:96]: warning that the dataloader only makes 1 fetch per request. when requested again, it'll use the cached version.
since you need to mutate the `voterIds` field, you either need to not use the dataloader, or mutate the cached value in addition to mutating the record in the DB.

e.g.
```js
reflectionGroups.forEach((group) => group.voterIds = [])
```

**Comment by mattkrick** [comment]
Screen Shot 2020-09-30 at 9 23 50 AM
here's how i debugged:
- was the bad value on client or server? i saw the server is still sending down the wrong number, so i know the problem is on the server

this led me to find the bug was in the server mutation (it was used a cached value). see other comment

File: packages/server/graphql/mutations/resetMeetingToStage.ts
Comment (tiffanyhan) [reply packages/server/graphql/mutations/resetMeetingToStage.ts:124]: true! out of curiosity and if anything comes top of mind, how would you handle `Property 'update' does not exist on type 'RetroMeetingMember'.`?

**Comment by tiffanyhan** [comment]
Thanks for your reviews @nickoferrall and @mattkrick !

**Pull request merged**
